src/main/scala/io/iohk/ethereum/vm/Stack.scala

```diff
@@ -0,0 +1,95 @@
+package io.iohk.ethereum.vm
+
+import io.iohk.ethereum.utils.EitherExtensions._
+
+object Stack {
+  /**
+    * Stack max size as defined in the YP (9.1)
+    */
+  val DefaultMaxSize = 1024
+
+  def empty(maxSize: Int = DefaultMaxSize): Stack =
+    new Stack(Vector(), maxSize)
+}
+
+//TODO: consider a List with head being top of the stack (DUP,SWAP go at most the depth of 16)
+/**
+  * Stack for the EVM. Instruction pop their arguments from it and push their results to it.
+  * The Stack handles overflow and underflow errors.
+  */
+class Stack private(private val underlying: Vector[DataWord], val maxSize: Int) {
```

Comment:
We decided we don't want classes like `Stack`, `Memory` or `Storage` to be case classes. The reason is that the implementation details of those classes are subject to change and therefore should be hidden. The rule of thumb is that case classes are used for data which is to be accessed directly from the class - here it's not the case. The few extra lines of code (equals, toString etc) required for the encapsulation are pretty standard and shouldn't be a concern. Does that sound reasonable?